[QoS] Verify the additional lossless queues and PGs in QoS test in dual ToR scenario

stephenxs opened the pull request:
Introduce a way to verify the additional lossless queues and PGs in the QoS test in the dual ToR scenario

1. Add a CLI option to specify whether it is to verify ports with or without additional lossless PGs and queues.
   It will collect all the dual ToR ports at the beginning of the test. If the option is on, the additional lossless PGs/queues will be verified in the corresponding tests
   The port information will always be collected regardless of whether the CLI option is fed because non-dual ToR ports must be used while the option is not fed, which means we need to understand the ports' role in any case.
   We use a Lua script to fetch such info because it is much faster than intensively using `sonic-db-cli`
2. Update the logic to fetch buffer profiles from BUFFER_QUEUE and BUFFER_PG table according to the CLI
3. Two additional profiles are introduced for verifying the additional lossless PGs/queues in XON and XOFF test if the CLI option is on
4. For the headroom pool test, all 4 DSCPs are passed to the PTF script as well as the CLI option. The additional DSCPs will be skipped by the PTF docker according to whether the ports are with additional lossless PGs/queues.
5. For the DSCP2queue mapping test, the CLI option is passed to the PTF script so that the latter is able to check the mapping according to the CLI option.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)


### Back port request
- [ ] 201911
- [x] 202012

### Approach
#### What is the motivation for this PR?

#### How did you do it?

#### How did you verify/test it?

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

keboliu commented:
I think we should ask the community to cherry-pick  https://github.com/Azure/sonic-mgmt/pull/1205 to master instead of creating this new PR?

stephenxs commented:
> I think we should ask the community to cherry-pick [Azure#1205](https://github.com/Azure/sonic-mgmt/pull/1205) to master instead of creating this new PR?

It can't be cleanly cherry-picked, which means another PR is required...

stephenxs commented:
Community PR [#1379](https://github.com/Azure/sonic-mgmt/pull/1379) opened.

stephenxs committed:
Provide a WA to verify the additional lossless queues and PGs in QoS test in dual ToR scenario

1. Add a CLI option to specify whether it is to verify ports with or without additional lossless PGs and queues.
   It will collect all the dual ToR ports at the beginning of the test. If the option is on, the additional lossless PGs/queues will be verified in the corresponding tests
2. Update the logic to fetch buffer profiles from BUFFER_QUEUE and BUFFER_PG table according to the CLI
2. Two additional profiles are introduced for verifying the additional lossless PGs/queues in XON and XOFF test if the CLI option is on
3. For headroom pool test, all 4 DSCPs are passed to PTF script as well as the CLI option. The additional DSCPs will be skipped by the PTF docker according to whether the ports are with additional lossless PGs/queues.
4. For DSCP2queue mapping test, the CLI option is passed to PTF script so that the latter is able to check the mapping according to the CLI option.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

stephenxs committed:
Update PG/queue

Signed-off-by: Stephen Sun <stephens@nvidia.com>

stephenxs committed:
Revert "Update PG/queue"

This reverts commit e9e93d79f07d4c3f67c83bf70f07d7704e4535f9.

stephenxs committed:
Fix lossless PG id

Signed-off-by: root <stephens@nvidia.com>

stephenxs committed:
Update according to different DSCP_TO_TC_MAP between uplinks/downlinks

Signed-off-by: Stephen Sun <stephens@nvidia.com>

stephenxs committed:
Use Lua plugin to accelerate the process to fetch dual ToR ports

Signed-off-by: Stephen Sun <stephens@nvidia.com>

stephenxs committed:
Remove unused code

Signed-off-by: Stephen Sun <stephens@nvidia.com>

stephenxs commented:
FYI.
Using `sonic-db-cli` to fetch ports' information: ~55 seconds
```
2022-07-08 06:51:19 qos_sai_base.initDualTorPorts            L0092 INFO   | Starting fetching dual ToR info
2022-07-08 06:52:14 qos_sai_base.initDualTorPorts            L0123 INFO   | Finish fetching dual ToR info set([u'Ethernet96', u'Ethernet124', u'Ethernet92', u'Ethernet120', u'Ethernet68', u'Ethernet76', u'Ethernet108', u'Ethernet88', u'Ethernet72', u'Ethernet64', u'Ethernet116', u'Ethernet100', u'Ethernet80', u'Ethernet112', u'Ethernet104', u'Ethernet84'])
```
Using Lua script to do it: 1 second
```
2022-07-08 08:23:54 qos_sai_base.initDualTorPorts            L0130 INFO   | Starting fetching dual ToR info
2022-07-08 08:23:55 qos_sai_base.initDualTorPorts            L0138 INFO   | Finish fetching dual ToR info set([u'Ethernet96', u'Ethernet124', u'Ethernet92', u'Ethernet120', u'Ethernet68', u'Ethernet76', u'Ethernet108', u'Ethernet88', u'Ethernet72', u'Ethernet64', u'Ethernet116', u'Ethernet100', u'Ethernet80', u'Ethernet112', u'Ethernet104', u'Ethernet84'])
```
